linux-syscall/src/time.rs

```diff
@@ -1,32 +1,199 @@
 
+/// TimeSpec struct for clock_gettime, similar to Timespec
 #[repr(C)]
 #[derive(Debug, Copy, Clone)]
 pub struct TimeSpec {
     /// seconds
-    sec: usize,
+    pub sec: usize,
     /// nano seconds
-    nsec: usize,
+    pub nsec: usize,
+}
+
+/// TimeVal struct for gettimeofday
+#[repr(C)]
+#[derive(Debug, Copy, Clone)]
+pub struct TimeVal {
+    /// seconds
+    sec: usize,
+    /// microsecond
+    usec: usize,
 }
 
 impl Syscall<'_> {
     /// finds the resolution (precision) of the specified clock clockid, and,
     /// if buffer is non-NULL, stores it in the struct timespec pointed to by buffer
     pub fn sys_clock_gettime(&self, clock: usize, mut buf: UserOutPtr<TimeSpec>) -> SysResult {
         info!("clock_gettime: id={:?} buf={:?}", clock, buf);
+        // TODO: handle clock_settime
+        let ts = TimeSpec::new();
+        buf.write(ts)?;
+
+        info!("TimeSpec: {:?}", ts);
+
+        Ok(0)
+    }
+
+    /// get the time with second and microseconds
+    pub fn sys_gettimeofday(
+        &mut self,
+        mut tv: UserOutPtr<TimeVal>,
+        tz: UserInPtr<u8>,
+    ) -> SysResult {
+        info!("gettimeofday: tv: {:?}, tz: {:?}", tv, tz);
+        // don't support tz
+        if !tz.is_null() {
+            return Err(LxError::EINVAL);
+        }
+
+        let timeval = TimeVal::new();
+        tv.write(timeval)?;
+
+        info!("TimeVal: {:?}", timeval);
+
+        Ok(0)
+    }
+
+    /// get time in seconds
+    #[cfg(target_arch = "x86_64")]
+    pub fn sys_time(&mut self, mut time: UserOutPtr<u64>) -> SysResult {
+        info!("time: time: {:?}", time);
+        let sec = TimeSpec::new().sec;
+        time.write(sec as u64)?;
+        Ok(sec)
+    }
+
+    /// get resource usage
+    /// currently only support ru_utime and ru_stime:
+    /// - `ru_utime`: user CPU time used
+    /// - `ru_stime`: system CPU time used
+    pub fn sys_getrusage(&mut self, who: usize, mut rusage: UserOutPtr<RUsage>) -> SysResult {
+        info!("getrusage: who: {}, rusage: {:?}", who, rusage);
+
+        let new_rusage = RUsage {
+            utime: TimeVal::new(),
+            stime: TimeVal::new(),
+        };
+        rusage.write(new_rusage)?;
+        Ok(0)
+    }
+
+    /// stores the current process times in the struct tms that buf points to
+    pub fn sys_times(&mut self, mut buf: UserOutPtr<Tms>) -> SysResult {
+        info!("times: buf: {:?}", buf);
+
+        let tv = TimeVal::new();
+
+        let tick = (tv.sec * 1_000_000 + tv.usec) / USEC_PER_TICK;
+
+        let new_buf = Tms {
+            tms_utime: 0,
+            tms_stime: 0,
+            tms_cutime: 0,
+            tms_cstime: 0,
+        };
+
+        buf.write(new_buf)?;
 
+        info!("tick: {:?}", tick);
+        Ok(tick as usize)
+    }
+}
+
+impl TimeVal {
+    /// create TimeVal
+    pub fn new() -> TimeVal {
+        TimeSpec::new().into()
+    }
+}
+
+impl TimeSpec {
+    /// create TimeSpec
+    pub fn new() -> TimeSpec {
```

Comment:
Maybe a better name: `now`.
Or `impl From<Duration> for TimeSpec`, then `TimeSpec::from(timer_now())`.
Which one is better?